### PR TITLE
[bugfix] Added the ability to associate subnets with a secondary vrf

### PR DESCRIPTION
**Pull request by shrsr**

### CODE REVIEW

File: aci/resource_aci_cloudsubnet.go [hunk 1]
```diff
@@ -89,10 +89,16 @@ func resourceAciCloudSubnet() *schema.Resource {
 			},
 
 			"relation_cloud_rs_subnet_to_flow_log": &schema.Schema{
-				Type: schema.TypeString,
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (akinross) [inline aci/resource_aci_cloudsubnet.go:93]: Is there a reason why this is not computed?

File: aci/resource_aci_cloudsubnet.go [hunk 2]
```diff
@@ -259,8 +270,16 @@ func resourceAciCloudSubnetCreate(ctx context.Context, d *schema.ResourceData, m
 		cloudSubnetAttr.SubnetGroup = subnetGroup.(string)
 	}
 
-	cloudSubnet, err := aciClient.CreateCloudSubnet(ip, CloudCIDRPoolDn, desc, cloudSubnetAttr, zoneDn)
+	var cloudRsSubnetToCtx string
+	if relationTocloudRsSubnetToCtx, ok := d.GetOk("relation_cloud_rs_subnet_to_ctx"); ok {
+		cloudRsSubnetToCtx = GetMOName(relationTocloudRsSubnetToCtx.(string))
+	} else {
+		cloudRsSubnetToCtx = ""
+	}
```
Comment (akinross) [inline aci/resource_aci_cloudsubnet.go:278]: do we need this else statement? is the var initialisation not already setting it to ""

File: aci/resource_aci_cloudsubnet.go [hunk 3]
```diff
@@ -259,8 +270,16 @@ func resourceAciCloudSubnetCreate(ctx context.Context, d *schema.ResourceData, m
...
+
+	cloudSubnet, err := aciClient.CreateCloudSubnet(ip, CloudCIDRPoolDn, desc, cloudSubnetAttr, zoneDn, cloudRsSubnetToCtx)
 	if err != nil {
+		log.Printf("HERE err %v", err)
```
Comment (akinross) [inline aci/resource_aci_cloudsubnet.go:282]: remove log

File: aci/resource_aci_cloudsubnet.go [hunk 4]
```diff
@@ -347,7 +371,14 @@ func resourceAciCloudSubnetUpdate(ctx context.Context, d *schema.ResourceData, m
 		cloudSubnetAttr.SubnetGroup = subnetGroup.(string)
 	}
 
-	cloudSubnet, err := aciClient.UpdateCloudSubnet(ip, CloudCIDRPoolDn, desc, cloudSubnetAttr, zoneDn)
+	var cloudRsSubnetToCtx string
+	if relationTocloudRsSubnetToCtx, ok := d.GetOk("relation_cloud_rs_subnet_to_ctx"); ok {
+		cloudRsSubnetToCtx = GetMOName(relationTocloudRsSubnetToCtx.(string))
+	} else {
+		cloudRsSubnetToCtx = ""
+	}
```
Comment (akinross) [inline aci/resource_aci_cloudsubnet.go:379]: do we need this else statement? is the var initialisation not already setting it to ""

**Review by akinross — CHANGES_REQUESTED** [review]

File: aci/resource_aci_cloudsubnet.go (hunk 2 above)
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go:278]: True. I removed it

File: aci/resource_aci_cloudsubnet.go (hunk 3 above)
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go:282]: ok

File: aci/resource_aci_cloudsubnet.go (hunk 1 above)
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go:93]: I added computed to it

File: aci/resource_aci_cloudsubnet.go (hunk 4 above)
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go:379]: True. I removed it

**Review by akinross — CHANGES_REQUESTED** [review]

**Review by anvitha-jain — DISMISSED** [review]
LGTM

File: aci/resource_aci_cloudsubnet.go [hunk 5]
```diff
@@ -89,10 +89,17 @@ func resourceAciCloudSubnet() *schema.Resource {
 			},
 
 			"relation_cloud_rs_subnet_to_flow_log": &schema.Schema{
-				Type: schema.TypeString,
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
 
+			"relation_cloud_rs_subnet_to_ctx": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go:97]: yes!

**Review by anvitha-jain — CHANGES_REQUESTED** [review]
relation_cloud_rs_subnet_to_ctx is a new parameter, do we need to update the docs for the aci_cloud_subnet

**Comment by shrsr** [comment]
> relation_cloud_rs_subnet_to_ctx is a new parameter, do we need to update the docs for the aci_cloud_subnet

Yes and done!

File: aci/data_source_aci_cloudsubnet.go [hunk 6]
```diff
@@ -55,15 +55,23 @@ func dataSourceAciCloudSubnet() *schema.Resource {
...
 				Description: "Only applicable to the AWS vendor",
 			},
 			"relation_cloud_rs_subnet_to_flow_log": &schema.Schema{
 				Type:     schema.TypeString,
 				Optional: true,
+				Computed: true,
+			},
+			"relation_cloud_rs_subnet_to_ctx": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (akinross) [inline aci/data_source_aci_cloudsubnet.go:68]: we don't need optional in the datasource I tthink for non-required attributes.

Comment (akinross) [reply aci/data_source_aci_cloudsubnet.go:68]: in docs you have set correctly but the entire schema needs to be adjusted

**Review by akinross — CHANGES_REQUESTED** [review]

**Review by akinross — DISMISSED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

File: aci/resource_aci_cloudsubnet.go [hunk 7]
```diff
@@ -329,6 +344,11 @@ func resourceAciCloudSubnetUpdate(ctx context.Context, d *schema.ResourceData, m
 		checkDns = append(checkDns, newRelParam.(string))
 	}
 
+	if d.HasChange("relation_cloud_rs_subnet_ctx") {
```
Comment (shrsr) [inline aci/resource_aci_cloudsubnet.go:347]: By @lhercot -> Change relationship name to the correct one

File: website/docs/r/cloud_subnet.html.markdown [hunk 8]
```diff
@@ -49,6 +49,7 @@ resource "aci_cloud_subnet" "foocloud_subnet" {
 * `usage` - (Optional) The usage of the port. This property shows how the port is used. Allowed values are "user", "gateway" and "infra-router". Default is "user". To make any subnet a Gateway subnet use `usage` = "gateway".	
 * `zone` - (Optional) [AWS Only] Availability zone where the subnet must be deployed. This property can carry both the actual zone or the ACI logical zone name. In the former case, the driver directly uses the value of this property. In the latter case, the Connector has to first resolve the mapping from ACI logical zone to the actual AWS zone. This parameter is required in APIC v5.0 or higher
 * `relation_cloud_rs_subnet_to_flow_log` - (Optional) Relation to class cloudAwsFlowLogPol. Cardinality - N_TO_ONE. Type - String.
+* `relation_cloud_rs_subnet_to_ctx` - (Optional) Relation to the class cloudRsSubnetToCtx. Type - String.
```
Comment (lhercot) [inline website/docs/r/cloud_subnet.html.markdown:52]:
```suggestion
* `relation_cloud_rs_subnet_to_ctx` - (Optional) Relation to associate the subnet with a secondary VRF (class cloudRsSubnetToCtx). Type - String.
```

**Review by lhercot — CHANGES_REQUESTED** [review]

File: website/docs/r/cloud_subnet.html.markdown (hunk 8 above)
Comment (shrsr) [reply website/docs/r/cloud_subnet.html.markdown]: done

File: aci/resource_aci_cloudsubnet.go (hunk 7 above)
Comment (shrsr) [reply aci/resource_aci_cloudsubnet.go]: done

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**